fastlane/lib/fastlane/actions/build_app.rb

```diff
@@ -1,13 +1,160 @@
+          # If that's the case, we won't set the provisioning profiles
+          # see https://github.com/fastlane/fastlane/issues/9490
+          if values[:export_options].kind_of?(Hash)
+            match_mapping = (Actions.lane_context[SharedValues::MATCH_PROVISIONING_PROFILE_MAPPING] || {}).dup
+            existing_mapping = (values[:export_options][:provisioningProfiles] || {}).dup
+
+            # Be smart about how we merge those mappings in case there are conflicts
+            mapping_object = Gym::CodeSigningMapping.new
+            hash_to_use = mapping_object.merge_profile_mapping(primary_mapping: existing_mapping,
+                                                             secondary_mapping: match_mapping,
+                                                                export_method: values[:export_method])
+
+            values[:export_options][:provisioningProfiles] = hash_to_use
+          else
+            self.show_xcode_9_warning
+          end
+        elsif Actions.lane_context[SharedValues::SIGH_PROFILE_PATHS]
+          # Since Xcode 9 you need to explicitly provide the provisioning profile per app target
+          # If the user used sigh we can match the profiles from sigh
+          values[:export_options] ||= {}
+          if values[:export_options].kind_of?(Hash)
+            # It's not always a hash, because the user might have passed a string path to a ready plist file
+            # If that's the case, we won't set the provisioning profiles
+            # see https://github.com/fastlane/fastlane/issues/9684
+            values[:export_options][:provisioningProfiles] ||= {}
+            Actions.lane_context[SharedValues::SIGH_PROFILE_PATHS].each do |profile_path|
+              begin
+                profile = FastlaneCore::ProvisioningProfile.parse(profile_path)
+                app_id_prefix = profile["ApplicationIdentifierPrefix"].first
+                bundle_id = profile["Entitlements"]["application-identifier"].gsub("#{app_id_prefix}.", "")
+                values[:export_options][:provisioningProfiles][bundle_id] = profile["Name"]
+              rescue => ex
+                UI.error("Couldn't load profile at path: #{profile_path}")
+                UI.error(ex)
+                UI.verbose(ex.backtrace.join("\n"))
+              end
+            end
+          else
+            self.show_xcode_9_warning
+          end
+        end
+
+        gym_output_path = Gym::Manager.new.work(values)
+        if gym_output_path.nil?
+          UI.important("No output path received from gym")
+          return nil
+        end
+
+        absolute_output_path = File.expand_path(gym_output_path)
+
+        absolute_dsym_path = absolute_output_path.gsub(/.ipa$/, ".app.dSYM.zip") if File.extname(absolute_output_path) == ".ipa"
+        absolute_dsym_path = absolute_output_path.gsub(/.pkg$/, ".dSYM.zip") if File.extname(absolute_output_path) == ".pkg"
```

Comment:
This file type check e.g. `File.extname(absolute_output_path) == ".ipa"` is done twice. Move these under
the check below.